### ISSUE TITLE
bug:helloworld exmaple failed

### ISSUE THREAD

**Issue by lordoffox** [alibaba/yalantinglibs#209]
### What happened + What you expected to happen
./helloworld_server
2023-02-16 06:39:09.036 INFO     [32259] [coro_rpc_server.hpp:317] begin to listen
2023-02-16 06:39:09.037 INFO     [32259] [coro_rpc_server.hpp:346] listen port 8801 successfully
2023-02-16 06:39:18.945 INFO     [32262] [coro_rpc_server.hpp:374] new client conn_id 1 coming
2023-02-16 06:39:18.946 ERROR    [32264] [coro_connection.hpp:150] bad magic number, conn_id 1
2023-02-16 06:49:08.150 INFO     [32262] [coro_rpc_server.hpp:374] new client conn_id 2 coming
2023-02-16 06:49:08.152 ERROR    [32265] [coro_connection.hpp:150] bad magic number, conn_id 2

./helloworld_client
2023-02-16 06:49:08.150 INFO     [32270] [coro_rpc_client.hpp:154] client_id 0 begin to connect 8801
2023-02-16 06:49:08.151 INFO     [32271] [coro_rpc_client.hpp:330] client_id 0 connect timer canceled, is_timeout_ 0, 0 , duration 5000 ms
2023-02-16 06:49:08.152 INFO     [32271] [coro_rpc_client.hpp:615] client_id 0 close
2023-02-16 06:49:08.152 INFO     [32271] [coro_rpc_client.hpp:330] client_id 0 rpc call timer canceled, is_timeout_ 0, 0 , duration 5000 ms
err: End of file
2023-02-16 06:49:08.152 ERROR    [32271] [coro_rpc_client.hpp:263] a closed client is not allowed call again, please create a new client
err: client has been closed, please create a new client
2023-02-16 06:49:08.152 ERROR    [32271] [coro_rpc_client.hpp:263] a closed client is not allowed call again, please create a new client
err: client has been closed, please create a new client
2023-02-16 06:49:08.152 ERROR    [32271] [coro_rpc_client.hpp:263] a closed client is not allowed call again, please create a new client
err: client has been closed, please create a new client
2023-02-16 06:49:08.152 ERROR    [32271] [coro_rpc_client.hpp:263] a closed client is not allowed call again, please create a new client
err: client has been closed, please create a new client
2023-02-16 06:49:08.152 ERROR    [32271] [coro_rpc_client.hpp:263] a closed client is not allowed call again, please create a new client
err: client has been closed, please create a new client
Done!

### Reproduction way
git clone https://github.com/alibaba/yalantinglibs.git
mkdir build && cd build
cmake ..
make -j
cd build/expamples
run helloworld_server helloworld_client

### Anything else
os:  ubuntu 22.10
gcc: 12

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

**Comment by poor-circle** [comment]
I can't reproduce it in my environment. Could you pull the code from main, delete cmake cache then try it again?

**Comment by lordoffox** [comment]
recompile have many error
/home/evan/yalantinglibs/include/struct_pack/struct_pack/struct_pack_impl.hpp:250:36: internal compiler error: in check_noexcept_r, at cp/except.cc:1074
  250 |       using T_ = decltype(get_types(T{}));
      |                           ~~~~~~~~~^~~~~
many gcc12 have some bug.

**Comment by poor-circle** [comment]
Could you show me your gcc version?
```
g++ --version
```

**Comment by lordoffox** [comment]
g++ --version
g++ (Ubuntu 12.2.0-3ubuntu1) 12.2.0

**Comment by lordoffox** [comment]
root@evansvr:/home/evan/yalantinglibs# rm -rf build/
root@evansvr:/home/evan/yalantinglibs# mkdir build && cd build
root@evansvr:/home/evan/yalantinglibs/build# cmake ..
-- The CXX compiler identification is GNU 12.2.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE  
-- CXX Standard: 20
-- Build type: Release
-- Build with libstdc++
CMake Warning at cmake/develop.cmake:34 (message):
  Sanitizer supported only for debug type
Call Stack (most recent call first):
  CMakeLists.txt:19 (include)


-- yLT_ENABLE_PROJECTS: struct_pack;struct_pb;struct_json;coro_rpc;easylog
-- ylt_add_tests_struct_pack
-- Found Protobuf: /usr/lib/x86_64-linux-gnu/libprotobuf.so (found version "3.12.4") 
-- ylt_add_benchmark_struct_pack
-- Found Boost: /usr/lib/x86_64-linux-gnu/cmake/Boost-1.74.0/BoostConfig.cmake (found version "1.74.0")  
-- Msgpack_FOUND: 1
-- ylt_add_example_struct_pack
-- ylt_add_tests_struct_pb
-- ylt_add_example_struct_pb
-- ylt_add_example_struct_json
-- ylt_add_tests_coro_rpc
-- ylt_add_benchmark_coro_rpc
-- GENERATE_BENCHMARK_DATA: ON
-- ylt_add_example_coro_rpc
-- ylt_add_tests_easylog
-- ylt_add_benchmark_easylog
-- Found Unwind: /usr/include/x86_64-linux-gnu (found suitable version "1.6.2", minimum required is "1.3.2") 
-- glog_FOUND: 1
-- Configuring done
-- Generating done
-- Build files have been written to: /home/evan/yalantinglibs/build

**Comment by lordoffox** [comment]
In file included from /home/evan/yalantinglibs/include/struct_pack/struct_pack.hpp:22,
                 from /home/evan/yalantinglibs/src/struct_pack/tests/test_serialize.cpp:31:
/home/evan/yalantinglibs/include/struct_pack/struct_pack/struct_pack_impl.hpp: In instantiation of ‘static constexpr bool struct_pack::detail::is_trivial_serializable<T>::solve() [with T = complicated_object]’:
/home/evan/yalantinglibs/include/struct_pack/struct_pack/struct_pack_impl.hpp:263:70:   required from ‘constexpr const bool struct_pack::detail::is_trivial_serializable<complicated_object>::value’
/home/evan/yalantinglibs/include/struct_pack/struct_pack/struct_pack_impl.hpp:659:49:   required from ‘consteval decltype(auto) struct_pack::detail::get_type_literal() [with Arg = complicated_object; ParentArgs = {nested_object}]’
/home/evan/yalantinglibs/include/struct_pack/struct_pack/struct_pack_impl.hpp:764:43:   required from ‘consteval decltype(auto) struct_pack::detail::get_types_literal_impl() [with Parent = nested_object; Args = {int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, person, complicated_object}]’
/home/evan/yalantinglibs/include/struct_pack/struct_pack/struct_pack_impl.hpp:774:61:   required from ‘consteval decltype(auto) struct_pack::detail::get_types_literal() [with T = nested_object; Args = {int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, person, complicated_object}]’
/home/evan/yalantinglibs/include/struct_pack/struct_pack/struct_pack_impl.hpp:888:74:   required from ‘consteval uint32_t struct_pack::detail::get_types_code_impl() [with T = nested_object; Args = {int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, person, complicated_object}; uint32_t = unsigned int]’
/home/evan/yalantinglibs/include/struct_pack/struct_pack/struct_pack_impl.hpp:894:67:   required from here
/home/evan/yalantinglibs/src/struct_pack/tests/test_serialize.cpp:531:51:   in ‘constexpr’ expansion of ‘struct_pack::get_type_code<nested_object>()’
/home/evan/yalantinglibs/include/struct_pack/struct_pack.hpp:146:73:   in ‘constexpr’ expansion of ‘struct_pack::detail::get_types_code<nested_object, std::tuple<int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, person, complicated_object> >()’
/home/evan/yalantinglibs/include/struct_pack/struct_pack/struct_pack_impl.hpp:1027:42:   in ‘constexpr’ expansion of ‘struct_pack::detail::get_types_code<nested_object, std::tuple<int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, person, complicated_object>, 0, 1, 2, 3>((std::make_index_sequence<4>(), std::make_index_sequence<4>()))’
/home/evan/yalantinglibs/include/struct_pack/struct_pack/struct_pack_impl.hpp:250:36: internal compiler error: in check_noexcept_r, at cp/except.cc:1074
  250 |       using T_ = decltype(get_types(T{}));
      |                           ~~~~~~~~~^~~~~
In file included from /home/evan/yalantinglibs/include/struct_pack/struct_pack.hpp:22,
                 from /home/evan/yalantinglibs/src/struct_pack/tests/test_stream.cpp:6:
/home/evan/yalantinglibs/include/struct_pack/struct_pack/struct_pack_impl.hpp: In instantiation of ‘static constexpr bool struct_pack::detail::is_trivial_serializable<T>::solve() [with T = nested_object]’:
/home/evan/yalantinglibs/include/struct_pack/struct_pack/struct_pack_impl.hpp:263:70:   required from ‘constexpr const bool struct_pack::detail::is_trivial_serializable<nested_object>::value’
/home/evan/yalantinglibs/include/struct_pack/struct_pack/struct_pack_impl.hpp:992:50:   required from ‘constexpr struct_pack::detail::size_info struct_pack::detail::calculate_one_size(const T&) [with T = nested_object]’
/home/evan/yalantinglibs/include/struct_pack/struct_pack/struct_pack_impl.hpp:1022:30:   required from here
/home/evan/yalantinglibs/src/struct_pack/tests/test_stream.cpp:38:52:   in ‘constexpr’ expansion of ‘struct_pack::get_needed_size<>(nested)’
/home/evan/yalantinglibs/include/struct_pack/struct_pack.hpp:173:50:   in ‘constexpr’ expansion of ‘struct_pack::detail::get_serialize_runtime_info<struct_pack::serialize_config{struct_pack::type_info_config::automatic}, nested_object>((* & args#0))’
/home/evan/yalantinglibs/include/struct_pack/struct_pack/struct_pack_impl.hpp:1300:40:   in ‘constexpr’ expansion of ‘struct_pack::detail::calculate_payload_size<nested_object>((* & args#0))’
/home/evan/yalantinglibs/include/struct_pack/struct_pack/struct_pack_impl.hpp:250:36: internal compiler error: in check_noexcept_r, at cp/except.cc:1074
  250 |       using T_ = decltype(get_types(T{}));
      |                           ~~~~~~~~~^~~~~
0x7f2eefc2350f __libc_start_call_main
        ../sysdeps/nptl/libc_start_call_main.h:58
0x7f2eefc235c8 __libc_start_main_impl
        ../csu/libc-start.c:381
Please submit a full bug report, with preprocessed source (by using -freport-bug).
Please include the complete backtrace with any bug report.
See <file:///usr/share/doc/gcc-12/README.Bugs> for instructions.
0x7f1320e2350f __libc_start_call_main
        ../sysdeps/nptl/libc_start_call_main.h:58
0x7f1320e235c8 __libc_start_main_impl
        ../csu/libc-start.c:381
Please submit a full bug report, with preprocessed source (by using -freport-bug).
Please include the complete backtrace with any bug report.
See <file:///usr/share/doc/gcc-12/README.Bugs> for instructions.
[ 48%] Linking CXX executable ../../../tests/test_util
make[2]: *** [src/struct_pack/tests/CMakeFiles/test_serialize.dir/build.make:76: src/struct_pack/tests/CMakeFiles/test_serialize.dir/test_serialize.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [src/struct_pack/tests/CMakeFiles/test_serialize.dir/build.make:188: src/struct_pack/tests/CMakeFiles/test_serialize.dir/test_stream.cpp.o] Error 1

**Comment by poor-circle** [comment]
I test it with gcc 11.3 in ubuntu 22.04, gcc12.2 & gcc 10.2 in centos, it work will.
maybe you can just run this command to skip test
```
cmake --build . --target helloworld_server
cmake --build . --target helloworld_client
cmake --build . --target helloworld_concurrency_client
```

**Comment by lordoffox** [comment]
./helloworld_client 
2023-02-16 11:16:36.094 INFO     [39376] [coro_rpc_client.hpp:158] client_id 0 begin to connect 8801
2023-02-16 11:16:36.199 INFO     [39377] [coro_rpc_client.hpp:617] client_id 0 close

ok.
it's work fine.
thanks you.

**Comment by lordoffox** [comment]
another quest.
server show a error:
2023-02-16 11:17:47.226 ERROR    [39373] [coro_connection.hpp:128] End of file, read head error
it's not an error when client disconnect.

**Comment by poor-circle** [comment]
> another quest. server show a error: 2023-02-16 11:17:47.226 ERROR [39373] [coro_connection.hpp:128] End of file, read head error it's not an error when client disconnect.

I will fix it later.

**Comment by poor-circle** [comment]
> another quest. server show a error: 2023-02-16 11:17:47.226 ERROR [39373] [coro_connection.hpp:128] End of file, read head error it's not an error when client disconnect.

Fixed by #205

**Comment by poor-circle** [comment]
I reproduce it in ubuntu 22.10 with g++ 12.2.0.
I will try to workaround this compile crash later.

**Comment by poor-circle** [comment]
Fixed by #213.
